### PR TITLE
fixes #151 coll.isEqualCollection compare 2 collections in any order

### LINKED ISSUE #151: Feature Request: expose isEqualCollection to cypher
See: [CollectionUtils#isEqualCollection](https://commons.apache.org/proper/commons-collections/apidocs/org/apache/commons/collections4/CollectionUtils.html#isEqualCollection%28java.util.Collection,%20java.util.Collection%29)

This would be very useful to check if two collections are the same. As far as I'm aware this isn't exposed to cypher.

### PR DESCRIPTION
**Pull request by vboulaye**
Fixes #151

This is an attempt at implementing the new function coll.isEqualCollection that returns true if the two collections contain the same elements with the same cardinality in any order

## Proposed Changes (Mandatory)

  - new UserFunction coll.isEqualCollection, internally uses an HashMap of object => cardinality to keep track of the cardinality of each element

### CODE REVIEW

**Comment by conker84** [comment]
@vboulaye can you push the `Coll` class with only your changes? It's difficult to review it with all the formatting stuff.
Thanks a lot.

**Comment by vboulaye** [comment]
sorry about that.
I didn't see it because I ignore all blanks in the diff viewer in intellij !

File: src/main/java/apoc/coll/Coll.java [hunk 1]
```diff
@@ -431,6 +431,30 @@ public boolean containsAllSorted(@Name("coll") List<Object> coll, @Name("values"
         return true;
     }
 
+    @UserFunction
+    @Description("apoc.coll.isEqualCollection(coll, values) return true if two collections contain the same elements with the same cardinality in any order (using a HashMap)")
```
Comment (conker84) [inline src/main/java/apoc/coll/Coll.java]: Can we simplify this? The implementation is quite complex, I think that something like this should do the work:
```
        Map<Object, Long> map1 = first.stream()
                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
        Map<Object, Long> map2 = second.stream()
                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
        return map1.equals(map2);
```

Comment (vboulaye) [reply src/main/java/apoc/coll/Coll.java]: this is simpler indeed!

File: src/main/java/apoc/coll/Coll.java [hunk 2]
```diff
@@ -431,6 +432,17 @@ public boolean containsAllSorted(@Name("coll") List<Object> coll, @Name("values"
...
+    public boolean isEqualCollection(@Name("coll") List<Object> first, @Name("values") List<Object> second) {
+        if (first == null || first.isEmpty() || first.size() != second.size()) return false;
```
Comment (conker84) [inline src/main/java/apoc/coll/Coll.java]: We must manage the null check for the second otherwise we get a NullPointerException at line 442. 
I think that we can also get rid of `first.size() != second.size()` because we have the count comparison for each collection element.

Please also add a

Comment (vboulaye) [reply src/main/java/apoc/coll/Coll.java]: I did not do the null check on second because it was not done on the existing function I started from `containsAll` . Should I heck it there too ?
I added some tests for the various null/empty values.
I kept the starting size check as it allows exiting early without having to create the intermediate Maps, but if you prefer I'll remove it.

Comment (conker84) [reply src/main/java/apoc/coll/Coll.java]: In that case makes sense because:
1. there is a null check on `coll`
2. the `call.containsAll` method provides a null check internally.
Please add the null check.

File: src/main/java/apoc/coll/Coll.java [hunk 3]
```diff
@@ -431,6 +432,18 @@ public boolean containsAllSorted(@Name("coll") List<Object> coll, @Name("values"
         return true;
     }
 
+    @UserFunction
+    @Description("apoc.coll.isEqualCollection(coll, values) return true if two collections contain the same elements with the same cardinality in any order (using a HashMap)")
+    public boolean isEqualCollection(@Name("coll") List<Object> first, @Name("values") List<Object> second) {
+        if (first == second) return true;
```
Comment (conker84) [inline src/main/java/apoc/coll/Coll.java]: @vboulaye please remove this line or use `equals` after the null check

**Comment by conker84** [comment]
@vboulaye please look at my last comment and we're done!
Thanks a lot for committing on this!

**Comment by conker84** [comment]
@vboulaye ping

File: src/main/java/apoc/coll/Coll.java (hunk 3 above)
Comment (vboulaye) [reply src/main/java/apoc/coll/Coll.java]: the problem is that if I remove this test, I have to add an explicit test to handle the case when both entries are null:  I believe  ̀apoc.coll.isEqualCollection(null,null)`  should return true .
I added 
```
        if (first == null && second == null) return true;
```
but I am not sure it is much better!

File: src/main/java/apoc/coll/Coll.java (hunk 2 above)
Comment (vboulaye) [reply src/main/java/apoc/coll/Coll.java]: I could not find the check on call.containsAll:I added a test and `apoc.coll.containsAll([1,2,3],null) ` throws a NullPointerException

File: src/main/java/apoc/coll/Coll.java (hunk 3 above)
Comment (vboulaye) [reply src/main/java/apoc/coll/Coll.java]: I just saw that in containsAllSorted the null check is done  with   if (coll == null || values == null) return false;
so containsAllSorted(null,null) is false

maybe this should be the same for  isEqualCollection ?

**Comment by conker84** [comment]
LGTM,
@vboulaye do you signed the CLA https://neo4j.com/developer/cla ?
